experimental/ixdtf/src/parser.rs

```diff
@@ -50,392 +53,532 @@ impl DecimalSeparator {
         }
-        return Ok(Some(year * sign));
     }
 
-    fn parse_date_four_digit_year<'a, I>(
-        iter: &mut core::iter::Peekable<I>,
-    ) -> Result<Option<i32>, ParseError>
-    where
-        I: Iterator<Item = &'a u8>,
-    {
-        let mut year: i32 = 0;
+    fn parse_date_four_digit_year(
+        s: &[u8],
+        start_index: usize,
+    ) -> Result<(Option<SegmentIndex>, &[u8]), ParseError> {
         let mut cnt = 0;
+        let mut mut_inner_remains = s;
         while cnt < 4 {
-            if let Some(y) = iter.peek() {
-                if **y >= b'0' && **y <= b'9' {
-                    year = year * 10 + **y as i32 - b'0' as i32;
+            if let Some((inner_first, inner_remains)) = mut_inner_remains.split_first() {
+                if (&b'0'..=&b'9').contains(&inner_first) {
+                    mut_inner_remains = inner_remains;
                 } else {
                     return Err(ParseError::DateFourDigitYear);
                 }
             } else {
                 return Err(ParseError::DateFourDigitYear);
             }
-            iter.next();
             cnt += 1;
         }
-        return Ok(Some(year));
+        return Ok((
+            Some(SegmentIndex {
+                start: start_index,
+                end: start_index + 3,
+            }),
+            mut_inner_remains,
+        ));
     }
 
-    fn parse_date_year<'a, I>(iter: &mut core::iter::Peekable<I>) -> Result<Option<i32>, ParseError>
-    where
-        I: Iterator<Item = &'a u8>,
-    {
-        if iter.peek() == Some(&&b'+') || iter.peek() == Some(&&b'-') {
-            return Self::parse_date_extended_year(iter);
-        }
-        return Self::parse_date_four_digit_year(iter);
-    }
-
-    fn parse_date_separator<'a, I>(
-        iter: &mut core::iter::Peekable<I>,
-        had_date_separator: &bool,
-    ) -> bool
-    where
-        I: Iterator<Item = &'a u8>,
-    {
-        // Whether current position has data separator.
-        let has_date_separator = iter.peek() == Some(&&b'-');
-        if *had_date_separator != has_date_separator {
-            return false;
-        }
-        if has_date_separator {
-            iter.next();
+    fn parse_date_year(
+        s: &[u8],
+        start_index: usize,
+    ) -> Result<(Option<SegmentIndex>, &[u8]), ParseError> {
+        if let Some((first, _)) = s.split_first() {
+            if first == &b'+' || first == &b'-' {
+                return Self::parse_date_extended_year(s, start_index);
+            }
+            return Self::parse_date_four_digit_year(s, start_index);
         }
-        return true;
+        return Err(ParseError::DateYear);
     }
 
-    fn parse_date_month<'a, I>(iter: &mut core::iter::Peekable<I>) -> Result<Option<u8>, ParseError>
-    where
-        I: Iterator<Item = &'a u8>,
-    {
+    fn parse_date_month(
+        s: &[u8],
+        start_index: usize,
+    ) -> Result<(Option<SegmentIndex>, &[u8]), ParseError> {
```

Comment:
```suggestion
    ) -> Result<u8, ParseError> {
```

Reply:
I prefer `Option<u8>` here because time could be optional.